src/sqlparser/src/ast/statement.rs

```diff
@@ -426,14 +427,21 @@ pub(super) fn fmt_create_items(
     } else {
         write!(&mut items, "{}", display_comma_separated(columns))?;
     }
-    if !columns.is_empty() && (!constraints.is_empty() || !watermarks.is_empty()) {
+    let mut leading_items = !columns.is_empty() || wildcard_idx.is_some();
+
+    if leading_items && !constraints.is_empty() {
         write!(&mut items, ", ")?;
     }
     write!(&mut items, "{}", display_comma_separated(constraints))?;
-    if !columns.is_empty() && !constraints.is_empty() && !watermarks.is_empty() {
+    leading_items |= !constraints.is_empty();
+
+    if leading_items && !watermarks.is_empty() {
```

Comment:
The logic is refactored a little bit to be cleaner and easier to reason about.

If patching in-place without refactor:
```diff
- !columns.is_empty() && (!constraints.is_empty() || !watermarks.is_empty())
+ (!columns.is_empty() && wildcard_idx.is_some()) && (!constraints.is_empty() || !watermarks.is_empty())

- !columns.is_empty() && !constraints.is_empty() && !watermarks.is_empty()
+ !constraints.is_empty() && !watermarks.is_empty()
```

Now let's reason about why fixing this way and how it is equivalent to the new code. Assume
* `A` = `!columns.is_empty() && wildcard_idx.is_some()`
* `B` = `!constraints.is_empty()`
* `C` = `!watermarks.is_empty()`

For the first bug, it is trivial that we shall fix `A` from `!columns.is_empty()` to include `&& wildcard_idx.is_some()`, because the wildcard is also part of the columns list.

For the second bug and the equivalence of refactoring, let's list down the conditions for the 2 commas:
* `A && (B || C)`
* `B && C` (previous buggy version `A && B && C`)

After refactor the conditions are:
* `A && B`
* `(A || B) && C`

You can also refer to the 2 new test cases for each bug correspondingly. The first test case is derived from a user-reported bug, while the second test case is constructed by comparing the difference between `A && B && C` and `B && C`.